### ISSUE TITLE
unsatisfied link error  /tmp/libpi4j.so on Raspbian Wheezy

### ISSUE THREAD

**Issue by savageautomate** [Pi4J/pi4j-v1#1]
ERROR:  The native library file [/tmp/libpi4j.so] could not be loaded due to an unsatisfied link error; /tmp/libpi4j.so: /tmp/libpi4j.so: cannot open shared object file: No such file or directory

If using Raspbian Wheezy and OpenJDK, it seems that the libpi4j.so native library is not getting written out to the "tmp" directory on application startup.

**Comment by savageautomate** [comment]
What's going on is that pi4j.jar depends on a native JNI library called "libpi4j.so".  Pi4J packages the libpi4j.so native library inside the JAR container as a resource file (under the "lib" folder) and on start up Pi4J is extracting this file and writing it to a system temp directory.  Then Pi4J instructs the JVM to load this native library from the temporary path.   All this logic is in this source file:  
https://github.com/Pi4J/pi4j/blob/master/pi4j-core/src/main/java/com/pi4j/util/NativeLibraryLoader.java

The following workaround may be applied to try and get around this defect.  Before using the logic mentioned above to extract the native library from the JAR package, Pi4J first attempts to load the native library using the "System.loadLibrary()" method.  This method will attempt to find the required native library on the JVM's library path.  So you could manually extract the libpi4j.so file from the JAR and copy it to a directory on you Pi and then in the java arguments manually specify the directory where the libpi4j.so is located as a java.library.path.   See example below:  

If you copy the libpi4j.so file to "/var/lib/java/libpi4j.so" then use the following argument in your java command when starting the program.

```
java -Djava.library.path=/var/lib/java
```

Java should now find the native library and never execute the extraction and tmp file stuff.

**Comment by savageautomate** [comment]
It turns out that this is an issue related to hard float versus soft float.  The Pi4J native library was compiled against the Debian "Wheezy" soft-float image and users attempting to run it on the Raspbian "Wheezy" hard-float image are encountering this issue.   

To get around this issue, you can checkout the Pi4J sources and build the projects including the pi4j-native project against your Pi hardware.  This will include the correct hard/soft float native library embedded inside the Pi4j-core JAR.      (mvn clean install -P native)

Investigating a fix for this issue now.

**Comment by savageautomate** [comment]
Fixed and released in latest 0.0.2-SNAPSHOT

**Comment by savageautomate** [comment]
A new 0.0.2-SNAPSHOT build has been deployed. 

A new defect was found in the last build related to this unsatisfied link error.  The latest wiringPi sources now include both a static and shared library in the build.  The libpi4j native lib was originally getting linked against the static wiringPi lib but now that both static and dynamic libs are available at linking time, the GCC linker opted to link against the  dynamic lib since it did not have any explicit library linking instructions defined in the libpi4j make file.  Modified the libpi4j make file to explicitly link against the static wiringPi library.

**Comment by gorstj** [comment]
Edit: Ignore this comment. The error went away when I sued SNAPSHOT 0.0.2 from your site in place of the copy Boiler-Control was using. They must be using an old/incorrect version. Sorry.

Are there any other causes for the following error?
Exception in thread "main" java.lang.UnsatisfiedLinkError: /tmp/libpi4j.so: /tmp/libpi4j.so: cannot open shared object file: No such file or directory.

I am running on Raspbian (28/10/2012) on a 512mb Pi  trying to run Boiler-Control (https://github.com/james-jaynne/boiler-control) which includes pi4j-core-0.0.2-SNAPSHOT.jar.

The 0.0.2 release should contain the fix for soft/hard float issues? I have tried running against 0.0.3 but it fails.

**Comment by savageautomate** [comment]
@gorstj 

I think this is related to this defect: https://github.com/Pi4J/pi4j/issues/5
The wiringPi library that is statically compiled inside Pi4J is out of date and thus not handling the 512mb model correctly. 

I will be working on a fix for this on Monday/Tuesday.  I'll post back here so you can test against the update.

**Comment by gorstj** [comment]
Thanks for your help.
I think this is the reason I get:

INFO: Library [pi4j] loaded successfully using embedded resource file: [jar:file:/home/pi/java/libs/pi4j-core-0.0.2-SNAPSHOT.jar.1!/lib/hard-float/libpi4j.so]
WARNING: wiringPi: Unable to determine board revision from "0"
 -> You may want to check:
 -> http://www.raspberrypi.org/phpBB3/viewtopic.php?p=184410#p184410
 -> Assuming a Rev 1 board

When you do the new version will it be both 0.0.2 or a new version of 0.0.3 as Boiler-Control does not seem to work at all with 0.0.3 (https://github.com/james-jaynne/boiler-control). Has that much changed between 0.0.2 and 0.0.3?

pi@raspberrypi ~/java $ sudo ./run.sh uk/co/jaynne/BoilerControl
Exception in thread "main" java.lang.NoClassDefFoundError: com/pi4j/io/gpio/GpioFactory
    at uk.co.jaynne.gpio.GpioControlPi4J.<init>(GpioControlPi4J.java:24)
    at uk.co.jaynne.gpio.GpioControlPi4J.<init>(GpioControlPi4J.java:16)
    at uk.co.jaynne.gpio.GpioControlPi4J$SingletonHolder.<clinit>(GpioControlPi4J.java:31)
    at uk.co.jaynne.gpio.GpioControlPi4J.getInstance(GpioControlPi4J.java:35)
    at uk.co.jaynne.ControlBroker.<init>(ControlBroker.java:25)
    at uk.co.jaynne.ControlBroker.<init>(ControlBroker.java:12)
    at uk.co.jaynne.ControlBroker$SingletonHolder.<clinit>(ControlBroker.java:34)
    at uk.co.jaynne.ControlBroker.getInstance(ControlBroker.java:38)
    at uk.co.jaynne.Scheduler.<init>(Scheduler.java:11)
    at uk.co.jaynne.BoilerControl.main(BoilerControl.java:18)
Caused by: java.lang.ClassNotFoundException: com.pi4j.io.gpio.GpioFactory
    at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:423)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:356)

**Comment by savageautomate** [comment]
It will most likely be a new build revision at 0.0.4.  There are some major API changes between 0.0.2 and 0.0.3 and are not directly API compatible.  0.0.4 will be based on 0.0.3 and should remain API compatible with 0.0.3.  Even though the API changed, the modifications to your code should be pretty simple.  Please see the latest samples on www.pi4j.com to see the differences.  This change was put in place to better prepare and support additional GPIO capabilities above and beyond the GPIO provided directly on the Raspberry Pi hardware.  I will be posting some articles in the coming weeks on my blog (http://www.savagehomeautomation.com) on GPIO expansion using Pi4J.

**Comment by gorstj** [comment]
Will I be able to copy the libpi4j.so from 0.0.4 to 0.0.2 (just decompress the jar file, change the file versions, then recompress) as libpi4j is just a version of wiringpi? This is a temporary measure until the developer moves from 0.0.2 to the updated version?

**Comment by savageautomate** [comment]
@gorstj

I think that will work.  I don't believe there are any deprecated or altered native API changes at this layer.
(Extract latest version of file libpi4j.so from pi4j-core-0.0.4-SNAPSHOT.jar of and replace this file inside the pi4j-core-0.0.2-SNAPSHOT.jar)

**Comment by savageautomate** [comment]
The new 0.0.4-SNAPSHOT build is available for testing. This build is compiled against the latest wiringPi sources.

You can download it from the Maven repository here:
https://oss.sonatype.org/content/groups/public/com/pi4j/pi4j-core/0.0.4-SNAPSHOT/

Please report success or any remaining open defects/issues.

**Comment by gorstj** [comment]
Quote: (Extract latest version of file libpi4j.so from pi4j-core-0.0.4-SNAPSHOT.jar of and replace this file inside the pi4j-core-0.0.2-SNAPSHOT.jar)

This does not work for me. I am hoping the write of boiler control will update to use 0.0.4.
It comes up with the error:
Exception in thread "main" java.lang.NoClassDefFoundError: com/pi4j/io/gpio/GpioFactory
    at uk.co.jaynne.gpio.GpioControlPi4J.<init>(GpioControlPi4J.java:24)
    at uk.co.jaynne.gpio.GpioControlPi4J.<init>(GpioControlPi4J.java:16)
    at uk.co.jaynne.gpio.GpioControlPi4J$SingletonHolder.<clinit>(GpioControlPi4J.java:31)
    at uk.co.jaynne.gpio.GpioControlPi4J.getInstance(GpioControlPi4J.java:35)
    at uk.co.jaynne.ControlBroker.<init>(ControlBroker.java:25)
    at uk.co.jaynne.ControlBroker.<init>(ControlBroker.java:12)
    at uk.co.jaynne.ControlBroker$SingletonHolder.<clinit>(ControlBroker.java:34)
    at uk.co.jaynne.ControlBroker.getInstance(ControlBroker.java:38)
    at uk.co.jaynne.Scheduler.<init>(Scheduler.java:11)
    at uk.co.jaynne.BoilerControl.main(BoilerControl.java:18)
Caused by: java.lang.ClassNotFoundException: com.pi4j.io.gpio.GpioFactory
    at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
    at java.net.URLClassLoader$1.run(URLClassLoader.java:355)
    at java.security.AccessController.doPrivileged(Native Method)
    at java.net.URLClassLoader.findClass(URLClassLoader.java:354)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:423)
    at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:356)
    ... 10 more

**Comment by savageautomate** [comment]
Well, I guess there must be some other intra-dependencies.  Moving to 0.0.4 should get things working for you.

**Comment by kevinhooke** [comment]
I just tried running an app using Pi4J 0.0.4 on the new Java SE 8 dev preview on hard float Raspbian Wheezy and I'm getting the same error:
Exception in thread "main" java.lang.UnsatisfiedLinkError: /tmp/libpi4j.so: /tmp/libpi4j.so: cannot open shared object file: No such file or directory.

If I follow your steps to manually copy the hard float .so to /var/lib/java and pass -Djava.library.path=/var/lib/java then it does work... seems like some issue still not being able to copy the .so to /tmp? I'll use this workaround for now, but just though I'd let you know...

**Comment by savageautomate** [comment]
Hi Kevin,

Yes, I ran into the same issue.  ISSUE: https://github.com/Pi4J/pi4j/issues/23

It turned out to be an internal routine used to attempt to identify if the system is soft-float or hard-float.  

This has been fixed in 0.0.5-SNAPSHOT.
Available via Sonatype OSS Maven Repo:
https://oss.sonatype.org/content/groups/public/com/pi4j/pi4j-core/

The updated version should work with no problems, but I also added new arguments that can override the internal detection in case this comes up again at least there will be an override option.  

-Dpi4j.armel - this argument will force Pi4J to load the soft float native library
-Dpi4j.armhf - this argument will force Pi4J to load the hard float native library
